src/cargo/core/source/source_id.rs

```diff
@@ -39,6 +39,10 @@ struct SourceIdInner {
     /// WARNING: this is not always set for alt-registries when the name is
     /// not known.
     name: Option<String>,
+    /// Name of the alt registry in the [registries] table.
+    /// WARNING: this is not always set for alt-registries when the name is
+    /// not known.
+    alt_registry_key: Option<String>,
```

Comment:
This seems to be very subtle in comparison to `name`, and I can't quite follow all the differences. I would expect a registry to have one name, so I'm not sure why there are two values here. Also, ideally `name` should be removed due to its unreliable behavior, so it seems like going in the wrong direction to add another similar field.

It looks like this is only used in one place for a diagnostic message. Would it be possible in that location to query the name from `SourceConfigMap`? Perhaps `id2name` could fetch the name?

Reply:
It's very subtle: `name` could be the key from either the `[source]` or `[registries]` table, and `id2name` contains the name from the `[source]` table. `alt_registry_key` represents the name from the `registries` table.

It's a bit unfortunate that we have both tables, since they're so similar. If we could go back in time, I'd say we should not add a `[registries]` table, and just use `[source]` for alt-registries.

Reply:
Can you extend the doc comment a little to discuss the differences? It would be helpful to be able to understand when one is set and not the other, or vice-versa, or have guidance to know when to use one vs the other.

I'd still like to move towards removing these fields, but I think that can be deferred to another day.